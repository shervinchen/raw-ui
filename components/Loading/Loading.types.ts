import { HTMLAttributes } from 'react';

export interface BaseLoadingProps {
  type?: 'dot' | 'spin';
  color?: string;
  className?: string;
  size?: number;
}

type NativeLoadingProps = Omit<
  HTMLAttributes<HTMLDivElement>,
  keyof BaseLoadingProps
>;

export type LoadingProps = BaseLoadingProps & NativeLoadingProps;
