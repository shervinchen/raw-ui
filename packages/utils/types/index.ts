export type DeepPartial<T> = {
  [P in keyof T]?: T[P] extends Record<string, any> ? DeepPartial<T[P]> : T[P];
};

export type Merge<T, P> = P & Omit<T, keyof P>;
