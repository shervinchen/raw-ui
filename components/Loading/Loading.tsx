import React, { FC, PropsWithChildren } from "react";
import classNames from "classnames";

import { useDotLoadingCSS, useSpinLoadingCSS } from "./Loading.styles";
import { LoadingProps } from "./Loading.types";

const DotLoading: FC<PropsWithChildren<Omit<LoadingProps, "type">>> = (
  props
) => {
  const { className, color, size, children, ...restProps } = props;
  const { className: resolveClassName, styles } = useDotLoadingCSS({
    color,
    size,
  });
  const classes = classNames("raw-dot-loading", className, resolveClassName);

  return (
    <div className={classes} {...restProps}>
      <span className="raw-dot-loading-inner">
        <i />
        <i />
        <i />
      </span>
      {children && <div className="raw-dot-loading-text">{children}</div>}
      {styles}
    </div>
  );
};

const SpinLoading: FC<PropsWithChildren<Omit<LoadingProps, "type">>> = (
  props
) => {
  const { className, color, size, children, ...restProps } = props;
  const { className: resolveClassName, styles } = useSpinLoadingCSS({
    color,
    size,
  });
  const classes = classNames("raw-spin-loading", className, resolveClassName);

  return (
    <div className={classes} {...restProps}>
      <div className="raw-spin-loading-wrapper">
        <div className="raw-spin-loading-inner">
          {[...new Array(12)].map((_, index) => (
            <span key={`raw-spin-${index}`} />
          ))}
        </div>
      </div>
      {children && (
        <div className="raw-spin-loading-text">
          {children}
        </div>
      )}
      {styles}
    </div>
  );
};

const Loading: FC<PropsWithChildren<LoadingProps>> = ({
  type = "dot",
  color = "#666",
  className = "",
  children,
  ...restProps
}) => {
  return type === "dot" ? (
    <DotLoading color={color} className={className} {...restProps}>
      {children}
    </DotLoading>
  ) : (
    <SpinLoading color={color} className={className} {...restProps}>
      {children}
    </SpinLoading>
  );
};

export default Loading;
