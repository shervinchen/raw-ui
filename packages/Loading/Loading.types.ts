import { ComponentPropsWithoutRef } from 'react';
import { Merge } from '../utils';

export type LoadingType = 'dot' | 'spin';

export interface BaseLoadingProps {
  type?: LoadingType;
  color?: string;
  className?: string;
  size?: number;
}

export type LoadingProps = Merge<
  ComponentPropsWithoutRef<'div'>,
  BaseLoadingProps
>;
