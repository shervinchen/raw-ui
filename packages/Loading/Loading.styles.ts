import css from 'styled-jsx/css';
import { LoadingProps } from './Loading.types';

export const useDotLoadingCSS = ({ size = 4, color }: LoadingProps) => {
  return css.resolve`
    .raw-dot-loading {
      display: inline-flex;
      align-items: center;
      flex-direction: column;
    }

    .raw-dot-loading :global(.raw-dot-loading-inner) {
      display: flex;
      justify-content: center;
      align-items: center;
      width: 100%;
      height: 100%;
      user-select: none;
    }

    .raw-dot-loading :global(.raw-dot-loading-text) {
      font-size: 14px;
      color: ${color};
      margin-top: 0.5em;
    }

    .raw-dot-loading :global(.raw-dot-loading-inner > i) {
      width: ${size}px;
      height: ${size}px;
      border-radius: 50%;
      background-color: ${color};
      margin: 0 ${size / 2}px;
      display: inline-block;
      animation: loading-blink 1.4s infinite both;
    }

    .raw-dot-loading :global(.raw-dot-loading-inner > i:nth-child(2)) {
      animation-delay: 0.2s;
    }

    .raw-dot-loading :global(.raw-dot-loading-inner > i:nth-child(3)) {
      animation-delay: 0.4s;
    }

    @keyframes loading-blink {
      0% {
        opacity: 0.2;
      }

      20% {
        opacity: 1;
      }

      100% {
        opacity: 0.2;
      }
    }
  `;
};

export const useSpinLoadingCSS = ({ size = 20, color }: LoadingProps) => {
  return css.resolve`
    .raw-spin-loading {
      display: block;
      box-sizing: border-box;
    }

    .raw-spin-loading :global(.raw-spin-loading-wrapper) {
      position: relative;
      left: 50%;
      top: 50%;
      width: ${size}px;
      height: ${size}px;
      transform: translateX(-50%);
    }

    .raw-spin-loading :global(.raw-spin-loading-inner) {
      position: absolute;
      left: 50%;
      top: 50%;
      width: 100%;
      height: 100%;
    }

    .raw-spin-loading :global(.raw-spin-loading-text) {
      font-size: 14px;
      color: ${color};
      margin-top: 0.5em;
    }

    .raw-spin-loading :global(.raw-spin-loading-inner > span) {
      position: absolute;
      top: -3.9%;
      width: 24%;
      height: 8%;
      left: -10%;
      border-radius: 6px;
      background-color: ${color};
      animation: spinner 1.2s linear 0s infinite normal none running;
    }

    .raw-spin-loading :global(.raw-spin-loading-inner > span:nth-child(1)) {
      animation-delay: -1.2s;
      transform: rotate(0deg) translate(146%);
    }

    .raw-spin-loading :global(.raw-spin-loading-inner > span:nth-child(2)) {
      animation-delay: -1.1s;
      transform: rotate(30deg) translate(146%);
    }

    .raw-spin-loading :global(.raw-spin-loading-inner > span:nth-child(3)) {
      animation-delay: -1s;
      transform: rotate(60deg) translate(146%);
    }

    .raw-spin-loading :global(.raw-spin-loading-inner > span:nth-child(4)) {
      animation-delay: -0.9s;
      transform: rotate(90deg) translate(146%);
    }

    .raw-spin-loading :global(.raw-spin-loading-inner > span:nth-child(5)) {
      animation-delay: -0.8s;
      transform: rotate(120deg) translate(146%);
    }

    .raw-spin-loading :global(.raw-spin-loading-inner > span:nth-child(6)) {
      animation-delay: -0.7s;
      transform: rotate(150deg) translate(146%);
    }

    .raw-spin-loading :global(.raw-spin-loading-inner > span:nth-child(7)) {
      animation-delay: -0.6s;
      transform: rotate(180deg) translate(146%);
    }

    .raw-spin-loading :global(.raw-spin-loading-inner > span:nth-child(8)) {
      animation-delay: -0.5s;
      transform: rotate(210deg) translate(146%);
    }

    .raw-spin-loading :global(.raw-spin-loading-inner > span:nth-child(9)) {
      animation-delay: -0.4s;
      transform: rotate(240deg) translate(146%);
    }

    .raw-spin-loading :global(.raw-spin-loading-inner > span:nth-child(10)) {
      animation-delay: -0.3s;
      transform: rotate(270deg) translate(146%);
    }

    .raw-spin-loading :global(.raw-spin-loading-inner > span:nth-child(11)) {
      animation-delay: -0.2s;
      transform: rotate(300deg) translate(146%);
    }

    .raw-spin-loading :global(.raw-spin-loading-inner > span:nth-child(12)) {
      animation-delay: -0.1s;
      transform: rotate(330deg) translate(146%);
    }

    @keyframes spinner {
      0% {
        opacity: 1;
      }
      100% {
        opacity: 0.15;
      }
    }
  `;
};
