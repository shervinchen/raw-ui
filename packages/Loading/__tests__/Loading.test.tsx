import React from 'react';
import { render, screen } from '@testing-library/react';
import Loading from '..';
import { LoadingType } from '../Loading.types';

describe('Loading', () => {
  test('should match the snapshot', () => {
    const { asFragment } = render(<Loading />);
    expect(asFragment()).toMatchSnapshot();
  });

  test('should support custom class name', () => {
    render(<Loading className="custom-loading" />);
    expect(screen.getByTestId('dotLoading')).toHaveClass('custom-loading');
  });

  ['dot', 'spin'].forEach((item: LoadingType) => {
    test(`should render ${item} type`, () => {
      render(<Loading type={item} />);
      expect(screen.getByTestId(`${item}Loading`)).toHaveClass(
        `raw-${item}-loading`
      );
    });
  });

  test('should render text with dot type', () => {
    render(<Loading>Loading</Loading>);
    expect(screen.getByText('Loading')).toBeTruthy();
  });

  test('should render text with spin type', () => {
    render(<Loading type="spin">Loading</Loading>);
    expect(screen.getByText('Loading')).toBeTruthy();
  });
});
