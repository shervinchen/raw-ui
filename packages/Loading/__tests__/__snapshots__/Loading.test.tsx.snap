// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Loading should match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="raw-dot-loading jsx-2553539443"
    data-testid="dotLoading"
  >
    <span
      class="raw-dot-loading-inner"
    >
      <i />
      <i />
      <i />
    </span>
  </div>
</DocumentFragment>
`;
